File: app/src/main/java/net/bloople/manga/MangosUrl.java
package net.bloople.manga;

import android.os.Parcel;
import android.os.Parcelable;

import com.bumptech.glide.load.model.GlideUrl;
import com.bumptech.glide.load.model.LazyHeaders;

import java.io.IOException;
import java.net.URL;
import java.net.URLConnection;

import okhttp3.Credentials;

public class MangosUrl implements Parcelable {
    private String url;
    private String username;
    private String password;

    public MangosUrl(String url, String username, String password) {
        this.url = url;
        this.username = username;
        this.password = password;
    }

    public MangosUrl withAppendedPath(String pathSegment) {
        return new MangosUrl(url + pathSegment, username, password);
    }

    public GlideUrl toGlideUrl() {
        if(username != null && password != null) {
            String credential = Credentials.basic(username, password);
            return new GlideUrl(url, new LazyHeaders.Builder().addHeader("Authorization", credential).build());
        }

        return new GlideUrl(url);
    }

    public URLConnection toUrlConnection() throws IOException {
        if(username != null && password != null) {
            String credential = Credentials.basic(username, password);
            URLConnection connection = new URL(url).openConnection();
            connection.addRequestProperty("Authorization", credential);
            return connection;
        }

        return new URL(url).openConnection();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.url);
        dest.writeString(this.username);
        dest.writeString(this.password);
    }

    public void readFromParcel(Parcel source) {
        this.url = source.readString();
        this.username = source.readString();
        this.password = source.readString();
    }

    protected MangosUrl(Parcel in) {
        this.url = in.readString();
        this.username = in.readString();
        this.password = in.readString();
    }

    public static final Parcelable.Creator<MangosUrl> CREATOR = new Parcelable.Creator<MangosUrl>() {
        @Override
        public MangosUrl createFromParcel(Parcel source) {
            return new MangosUrl(source);
        }

        @Override
        public MangosUrl[] newArray(int size) {
            return new MangosUrl[size];
        }
    };
}

File: app/src/main/java/net/bloople/manga/MangosUrl.kt
package net.bloople.manga

import android.net.Uri
import android.os.Parcelable
import android.os.Parcel
import android.os.Parcelable.Creator
import android.util.Base64
import coil3.network.NetworkHeaders
import coil3.network.httpHeaders
import coil3.request.ImageRequest
import okhttp3.Request

open class MangosUrl(private val url: String, private val credential: String? = null) : Parcelable {
    constructor(url: String, username: String?, password: String?) : this(
        url,
        if(username != null && password != null) {
            Base64.encodeToString("$username:$password".toByteArray(), Base64.NO_WRAP)
        }
        else {
            null
        }
    )

    operator fun div(other: String): MangosUrl {
        return MangosUrl(url + "/" + Uri.encode(other), credential)
    }

    fun toOkHttpRequest(): Request {
        val builder = Request.Builder().url(url)
        if(credential != null) builder.header("Authorization", "Basic $credential")
        return builder.build()
    }

    fun loadInto(builder: ImageRequest.Builder): ImageRequest.Builder {
        return builder.apply {
            data(url)
            if(credential != null) {
                httpHeaders(NetworkHeaders.Builder().add("Authorization", "Basic $credential").build())
            }
        }
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(url)
        dest.writeString(credential)
    }

    protected constructor(input: Parcel) : this(input.readString()!!, input.readString())

    companion object {
        @JvmField
        val CREATOR: Creator<MangosUrl> = object : Creator<MangosUrl> {
            override fun createFromParcel(source: Parcel): MangosUrl {
                return MangosUrl(source)
            }

            override fun newArray(size: Int): Array<MangosUrl?> {
                return arrayOfNulls(size)
            }
        }
    }
}